src/nvim/msgpack_rpc/channel.c

```diff
@@ -442,17 +442,22 @@ static void handle_request(Channel *channel, msgpack_object *request)
 
   if (error.set) {
     // Validation failed, send response with error
-    channel_write(channel,
-                  serialize_response(channel->id,
-                                     request_id,
-                                     &error,
-                                     NIL,
-                                     &out_buffer));
-    char buf[256];
-    snprintf(buf, sizeof(buf),
-             "Channel %" PRIu64 " sent an invalid message, closing.",
-             channel->id);
-    call_set_error(channel, buf);
+    bool success = channel_write(channel,
+                                 serialize_response(channel->id,
+                                                    request_id,
+                                                    &error,
+                                                    NIL,
+                                                    &out_buffer));
+
+    // Checking the result of the above write because the channel may
+    // be closed and freed.
```

Comment:
I've swapped out the `success` boolean for the `if(channel_write(` approach you mentioned. I think it looks much clearer now.